source/server/listener_manager_impl.cc

```diff
@@ -832,11 +838,17 @@ std::unique_ptr<Network::FilterChain> ListenerFilterChainFactoryBuilder::buildFi
 
   std::vector<std::string> server_names(filter_chain.filter_chain_match().server_names().begin(),
                                         filter_chain.filter_chain_match().server_names().end());
-  return std::make_unique<FilterChainImpl>(
-      config_factory.createTransportSocketFactory(*message, factory_context_,
-                                                  std::move(server_names)),
-      listener_component_factory_.createNetworkFilterFactoryList(filter_chain.filters(),
-                                                                 filter_chain_factory_context));
+
+  auto filter_chain_res =
+      std::make_unique<FilterChainImpl>(config_factory.createTransportSocketFactory(
+                                            *message, factory_context_, std::move(server_names)),
+                                        listener_component_factory_.createNetworkFilterFactoryList(
+                                            filter_chain.filters(), *filter_chain_factory_context));
+  // TODO(lambdai): remove this ugly downcast
```

Comment:
Why is this needed?

Reply:
Ooops, i think it's not required. Let me check

Reply:
I don't have an immediate solution on fixing this...
The background is that we only have one implementation at downstream(listener, filter chain manager, network filter chain) but we do have network filter chain at upstream. I am adding downstream filter chains the drainable functionality per downstream filter chain context so it rely on implementation not the filter chain context common interface.

Meanwhile it's safe at this moment because ListenerManager knows exactly what it is doing

Reply:
This is fragile let's please fix this now. You might need to somehow split the upstream/downstream filter chain interfaces. If so, please do that in a separate PR first.